File: Practica5/QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;

/**
 *
 * @author Lorena
 */
public class Jugador {

    private boolean encarcelado;
    private String nombre;
    private int saldo;
    private Casilla casillaActual;
    private Sorpresa cartaLibertad;
    private ArrayList<TituloPropiedad> propiedades;
    static final int FactorEspeculador = 1;

    //Constructores
    public Jugador(String nombre) {
        this.nombre = nombre;
        encarcelado = false;
        saldo = 0;
        casillaActual = null;
        propiedades = new ArrayList();
    }

    Jugador() {

    }

    //Constructor de copia
    protected Jugador(Jugador jugador) {
        nombre = jugador.nombre;
        encarcelado = jugador.encarcelado;
        saldo = jugador.saldo;
        casillaActual = jugador.casillaActual;
        propiedades = jugador.propiedades;
    }

    //Consultores y modificadores
    public String getNombre() {
        return nombre;
    }

    public int getSaldo() {
        return saldo;
    }

    public static int getFactorEspeculador() {
        return FactorEspeculador;
    }

    public Casilla getCasillaActual() {
        return casillaActual;
    }

    public ArrayList<TituloPropiedad> getPropiedades() {
        return propiedades;
    }

    public boolean getEncarcelado() {
        return encarcelado;
    }

    void setCartaLibertad(Sorpresa carta) {
        cartaLibertad = carta;
    }

    void setCasillaActual(Casilla casilla) {
        casillaActual = casilla;
    }

    public void setEncarcelado(boolean encarcelado) {
        System.out.println(encarcelado);
        this.encarcelado = encarcelado;
    }

//    public static void setFactorEspeculador(int FactorEspeculador) {
//        Jugador.FactorEspeculador = FactorEspeculador;
//    }
    
    
    //Métodos
    public boolean tengoPropiedades() {
        if (propiedades.isEmpty()) {
            return false;
        } else {
            return true;
        }
    }

    protected boolean actualizarPosicion(Casilla casilla) {

        if (casilla.getNumeroCasilla() < casillaActual.getNumeroCasilla()) {
            int cantidad = Qytetet.SALIDA;
            this.modificarSaldo(cantidad);
        }
        boolean tienePropietario = false;
        this.setCasillaActual(casilla);

        if (casilla.soyEdificable()) {
            Calle aux = (Calle) casilla;
            if (aux.tengoPropietario()) {
                encarcelado = aux.propietarioEncarcelado();
                if (!encarcelado) {
                    int costeAlquiler = aux.cobrarAlquiler();
                    this.modificarSaldo(-costeAlquiler);
                }
                tienePropietario = true;
            }

        } else if (casilla.getTipo() == TipoCasilla.IMPUESTO) {
            Calle aux = (Calle) casilla;
            int coste = aux.getCoste();
            pagarImpuesto(coste);
        }

        return tienePropietario;
    }

    boolean comprarTitulo() {
        boolean puedoComprar = false;
        if (casillaActual.soyEdificable()) {
            Calle aux = (Calle) casillaActual;
            boolean tengoPropietario = aux.tengoPropietario();

            if (!tengoPropietario) {
                int costeCompra = aux.getCoste();

                if (costeCompra <= saldo) {
                    TituloPropiedad titulo = aux.asignarPropietario(this);
                    propiedades.add(titulo);
                    int cantidad = -costeCompra;
                    this.modificarSaldo(cantidad);
                    puedoComprar = true;
                }
            }
        }
        return puedoComprar;
    }

    //Preguntar
    Sorpresa devolverCartaLibertad() {
        Sorpresa devolver = null;
        if (tengoCartaLibertad()) {
            devolver = null;
            cartaLibertad = devolver;
        }
        return cartaLibertad;
    }

    void irACarcel(Casilla casilla) {
        this.setCasillaActual(casilla);
        this.setEncarcelado(true);
    }

    void modificarSaldo(int cantidad) {
        saldo += cantidad;
    }

    int obtenerCapital() {
        int capital = 0;
        int prop = 0;

        for (int i = 0; i < propiedades.size(); i++) {
            Calle aux = (Calle) propiedades.get(i).casilla;
            if (!propiedades.get(i).getHipotecada()) {
                prop += aux.getCoste() + this.cuantasCasasHotelesTengo() + propiedades.get(i).getPrecioEdificar();
            } else {
                prop += aux.getCoste() + this.cuantasCasasHotelesTengo() + propiedades.get(i).getPrecioEdificar()
                        - propiedades.get(i).getHipotecaBase();
            }
        }
        capital = saldo + prop;
        return capital;
    }

    ArrayList<TituloPropiedad> obtenerPropiedadesHipotecadas(boolean hipotecada) {
        ArrayList<TituloPropiedad> estanHipotecados = new ArrayList();
        for (int i = 0; i < propiedades.size(); i++) {
            if (propiedades.get(i).getHipotecada() == hipotecada) {
                estanHipotecados.add(propiedades.get(i));
            }
        }
        return estanHipotecados;
    }

    void pagarCobrarPorCasaYHotel(int cantidad) {
        int numeroTotal = this.cuantasCasasHotelesTengo();
        this.modificarSaldo(numeroTotal * cantidad);
    }

    boolean pagarLibertad(int cantidad) {
        boolean tengoSaldo = this.tengoSaldo(cantidad);

        if (tengoSaldo) {
            this.modificarSaldo(cantidad);
        }
        return tengoSaldo;
    }

    boolean puedoEdificarCasa(Calle casilla) {
        boolean esMia = this.esDeMiPropiedad(casilla);
        boolean tengoSaldo = false;
        if (esMia) {
            int cantidad = casilla.getPrecioEdificar();
            tengoSaldo = this.tengoSaldo(cantidad);
        }
        return tengoSaldo;
    }

    boolean puedoEdificarHotel(Calle casilla) {
        boolean esMia = this.esDeMiPropiedad(casilla);
        boolean tengoSaldo = false;
        if (esMia) {
            int cantidad = casilla.getPrecioEdificar();
            tengoSaldo = this.tengoSaldo(cantidad);
        }
        return tengoSaldo;
    }

    boolean puedoHipotecar(Calle casilla) {
        return this.esDeMiPropiedad(casilla);
    }

    boolean puedoPagarHipoteca(Casilla casilla) {
        throw new UnsupportedOperationException("Sin implementar");
    }

    boolean puedoVenderPropiedad(Calle casilla) {
        boolean hipotecada = casilla.estaHipotecada();
        if (!hipotecada && esDeMiPropiedad(casilla)) {
            return true;
        } else {
            return false;
        }
    }

    boolean tengoCartaLibertad() {
        if (cartaLibertad != null) {
            return true;
        } else {
            return false;
        }
    }

    void venderPropiedad(Casilla casilla) {
        Calle aux = (Calle) casilla;
        int precioVenta = aux.venderTitulo();
        this.modificarSaldo(precioVenta);
        this.eliminarDeMisPropiedades(casilla);
    }

    private int cuantasCasasHotelesTengo() {
        int total = 0;
        for (int i = 0; i < propiedades.size(); i++) {
            Calle aux = (Calle) propiedades.get(i).casilla;
            total += aux.getNumCasas() + aux.getNumHoteles();
        }
        return total;
    }

    private void eliminarDeMisPropiedades(Casilla casilla) {
        Calle aux = (Calle) casilla;
        propiedades.remove(aux.getTitulo());
    }

    public boolean esDeMiPropiedad(Calle casilla) { //??
        boolean esMia = false;
        for (int i = 0; i < propiedades.size(); i++) {
            if (propiedades.get(i) == casilla.getTitulo()) {
                esMia = true;
            }
        }
        return esMia;
    }

    private boolean tengoSaldo(int cantidad) {
        if (saldo >= cantidad) {
            return true;
        } else {
            return false;
        }
    }

    protected void pagarImpuesto(int cantidad) {
        modificarSaldo(-cantidad);
    }

    protected Especulador convertirme(int fianza) {
        Especulador especulador = new Especulador(this, fianza);
        // this.setFactorEspeculador(especulador.getFactorEspeculador());
        return especulador;
    }

    //Método toString() devuelve un String con el estado del objeto correspondiente
    @Override
    public String toString() {
        return "Nombre: " + nombre + "\n"
                + "Está encarcelado? " + encarcelado + "\n"
                + "Saldo: " + saldo + "\n"
                //+ "Casilla actual: " + casillaActual + "\n"
                + "Tiene carta libertad? " + cartaLibertad + "\n"
                + "Propiedades: " + propiedades.toString() + "\n";
                //+ "Factor especulador: " + FactorEspeculador + "\n";
    }

}

File: Practica4/QytetetRuby P4/lib/jugador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'sorpresa'
require_relative 'titulo_propiedad'
require_relative 'casilla'
require_relative 'calle'



module ModeloQytetet
  class Jugador
    @@FACTOR_ESPECULADOR = 1
    
    #Consultores y modificadores
    attr_accessor :encarcelado , :carta_libertad, :casilla_actual , :propiedades ,:carta_libertad , :FACTOR_ESPECULADOR
    attr_reader :nombre, :saldo
    
    #Constructor
#    def initialize(nom)
#      @nombre = nom
#      @encarcelado = false
#      @saldo = 0
#      @propiedades = Array.new
#      @casilla_actual = nil
#      @carta_libertad = nil
#    end
 #  @propiedades = Array.new
     def initialize(nom,encarcelado,saldo,casilla_a,carta_l,propiedades)
      @nombre = nom
      @encarcelado = encarcelado
      @saldo = saldo
      @casilla_actual = casilla_a
      @carta_libertad = carta_l
      @propiedades = propiedades
      
    end
    
    def self.nuevo_jugador(nom)
      Jugador.new(nom,false,0,nil,nil,Array.new)
    end
    
    def self.copia_jugador(jugador)
#      puts jugador
      Jugador.new(jugador.nombre,jugador.encarcelado,jugador.saldo,jugador.casilla_actual,jugador.carta_libertad,jugador.propiedades)
    end
    
    
    
    #Métodos
 
    
    def tengo_propiedades()
      if @propiedades.empty?
        return false
      else
        return true
      end
    end
    
    def actualizar_posicion(casilla) 
      if casilla.numero_casilla < @casilla_actual.numero_casilla
        modificar_saldo(1000)
      end
      
      tiene_propietario = false
      @casilla_actual = casilla
      
      if casilla.soy_edificable()
        if casilla.titulo.tengo_propietario()
          @encarcelado = casilla.propietario_encarcelado()
          if !@encarcelado
            coste_alquiler = casilla.cobrar_alquiler()
            modificar_saldo(-coste_alquiler)
          end
          tiene_propietario = true
        end   
      elsif casilla.tipo == TipoCasilla::IMPUESTO
        coste = casilla.coste
        modificar_saldo(-coste)
      
      end 
      tiene_propietario
    end
    
    def comprar_titulo()
      puedo_comprar = false
      if @casilla_actual.soy_edificable()
        tengo_propietario = @casilla_actual.tengo_propiedad()
        if !tengo_propietario
          coste_compra = @casilla_actual.coste
          
          if coste_compra <= @saldo
            titulo = @casilla_actual.asignar_propietario(self)
            @propiedades << titulo
            modificar_saldo(-coste_compra)
            puedo_comprar = true
          end
        end
      end
      puedo_comprar
    end
    
    def devolver_carta_libertad()
      devolver = @carta_libertad
      if(tengo_carta_libertad)
        @carta_libertad = nil
        @encarcleado = false
      end
      devolver
    end
    
    def ir_a_carcel(casilla)
      @casilla_actual = casilla 
      @encarcelado = true
    end
    
    def modificar_saldo(cantidad)
      @saldo += cantidad
    end
    
    def obtener_capital()
      @propiedades.each do |i|
        if !@propiedades.at(i).hipotecada
          prop += @propiedades.at(i).casilla.coste + cuantas_casas_hoteles_tengo + @propiedades.at(i).precio_edificar
        else
          prop += @propiedades.at(i).casilla.coste + cuantas_casas_hoteles_tengo + @propiedades.at(i).precio_edificar 
          - @propiedades.at(i).hipoteca_base
 
        end
      end
      capital = @saldo + prop
      capital
    end
    
    def obtener_propiedades_hipotecadas(esta_hipotecada)
      estan_hipotecados = Array.new
      for i in 0..@propiedades.size()-1
        if @propiedades.at(i).hipotecada == esta_hipotecada
          estan_hipotecados << @propiedades.at(i)
        end
      end
      estan_hipotecados
    end
    
    def pagar_cobrar_por_casa_y_hotel(cantidad)
      numero_total = cuantas_casas_hoteles_tengo()
      modificar_saldo(numero_total * cantidad)
    end
    
    def pagar_libertad(cantidad)
      tengo_saldo = tengo_saldo(cantidad)
      
      if tengo_saldo
        modificar_saldo(cantidad)
      end
      
      tengo_saldo
    end

    def puedo_edificar_casa(casilla)
      esMia = es_de_mi_propiedad(casilla)
      if(esMia)
        cantidad = casilla.get_precio
        tengo_saldo = tengo_saldo(cantidad)
      end
      return tengo_saldo
    end
    
    def puedo_edificar_hotel(casilla)
      esMia = es_de_mi_propiedad(casilla)
      if(esMia)
        cantidad = casilla.titulo.precio_edificar
        tengo_saldo = tengo_saldo(cantidad)
      end
      return tengo_saldo
    end
    
    def puedo_hipotecar(casilla)
      return es_de_mi_propiedad(casilla)
    end
    
    def puedo_pagar_hipoteca(casilla)
      
    end
    
    def puedo_vender_propiedad(casilla)
      esta_hipotecada = casilla.esta_hipotecada
      if(!esta_hipotecada && es_de_mi_propiedad(casilla))
        true
      else 
        false
      end
    end
    
    def tengo_carta_libertad()
      if(@cartaLibertad != nil)
        true
      else
        false
      end
    end
    
    def vender_propiedad(casilla)
      precio_venta = casilla.vender_titulo()
      modificar_saldo(precio_venta)
      eliminar_de_mis_propiedades(casilla)
    end
    
    def cuantas_casas_hoteles_tengo()
      @propiedades.each do |i|
        total += @propiedades.at(i).casilla.num_casas + @propiedades.at(i).casilla.num_hoteles
      end 
      total
    end
    
    def eliminar_de_mis_propiedades(casilla)
      @propiedades.delete(casilla.titulo)
    end
    
    def es_de_mi_propiedad(casilla)
      esmia = false
      for i in 0..@propiedades.size()-1
        if @propiedades.at(i) == casilla.titulo
          esmia=true         
        end
      end
      esmia
    end
    
    def tengo_saldo(cantidad)
      if @saldo >= cantidad
        true
      else
        false
      end
    end
    
    def pagar_impuestos(cantidad)
      modificar_saldo(-cantidad)
    end
    
    def convertirme(fianza)      
      espe = Especulador.new(self,fianza)
      espe
    end
    
    def get_factor_especulacion()
      @@FACTOR_ESPECULADOR
    end
    
    #Método toString() devuelve un String con el estado del objeto correspondiente
    def to_s
      "************************************************\n
       Nombre:  #{@nombre} \n
       Esta encarcelado? #{@encarcelado} \n
       Saldo:  #{@saldo} \n
       Tiene carta libertad? #{@carta_libertad} \n
       Propiedades: #{@propiedades} \n"
       #Factor especulación: #{@factor_especulador} \n"
       
    end
    
    private :cuantas_casas_hoteles_tengo, :eliminar_de_mis_propiedades, :tengo_saldo
    
  end
end
